app/components/metatip/ally.element.js

```diff
@@ -1,22 +1,81 @@
+import $ from 'blingblingjs'
 import { Metatip } from './metatip.element.js'
+import { TinyColor } from '@ctrl/tinycolor'
+import { getStyle, getComputedBackgroundColor } from '../../utilities'
+import { getContrastingColor } from '../../utilities'
+
+const modemap = {
+  'hex': 'toHexString',
+  'hsl': 'toHslString',
+  'rgb': 'toRgbString',
+}
 
 export class Ally extends Metatip {
   constructor() {
     super()
   }
 
+  copyToClipboard(text) {
+    navigator.clipboard.writeText(text)
+  }
+
+  copyColorSwatch(event) {
+    this.copyToClipboard(event.target.querySelector('span').innerText.trim())
+  }
+
+  observe() {
+    $('[color-swatch], [color-swatch] *', this.$shadow).on('click', this.copyColorSwatch.bind(this))
+  }
+
+  unobserve() {
+    $('[color-swatch], [color-swatch] *', this.$shadow).off('click', this.copyColorSwatch.bind(this))
+  }
+
   render({el, ally_attributes, contrast_results}) {
+    const colormode = modemap[$('vis-bug').attr('color-mode')]
+
+    const foreground = el instanceof SVGElement
+      ? (getStyle(el, 'fill') || getStyle(el, 'stroke'))
+      : getStyle(el, 'color')
+    const background = getComputedBackgroundColor(el)
+
+    const contrastingForegroundColor = getContrastingColor(foreground)
+    const contrastingBackgroundColor = getContrastingColor(background)
+
+    this.style.setProperty('--copy-message-left-color', contrastingForegroundColor)
+    this.style.setProperty('--copy-message-right-color', contrastingBackgroundColor)
+
     return `
       <figure>
         <header>
           <h5>${el.nodeName.toLowerCase()}${el.id && '#' + el.id}</h5>
         </header>
+        <div color-swatches>
+          <span color-swatch style="background-color:${foreground};">
+            <small style="color:${contrastingForegroundColor};">
+              Foreground
+            </small>
+            <span style="color:${contrastingForegroundColor};">
+              ${new TinyColor(foreground)[colormode]()}
+            </span>
+          </span>
+          <span color-swatch style="background-color:${background};">
+            <small style="color:${contrastingBackgroundColor};">
+              Background
+            </small>
+            <span style="color:${contrastingBackgroundColor};">
+              ${new TinyColor(background)[colormode]()}
+            </span>
+          </span>
+        </div>
         <code accessibility>
-          ${ally_attributes.reduce((items, attr) => `
-            ${items}
-            <span prop>${attr.prop}:</span>
-            <span value>${attr.value}</span>
-          `, '')}
+          <div>
```

Comment:
there's an empty div getting made here if not hovering something with aria roles or other screen reader information. try hovering on some of the form inputs on the demo page.

Reply:
i made the empty div disappear now - and the 10px gap in the actual presentation